<a href="https://colab.research.google.com/github/uisha/civi100-project/blob/main/CIVI100.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Dataset Description

**Source:** https://link.springer.com/chapter/10.1007/978-3-319-46466-4_13

## Overview
The **SD-198** dataset is a large-scale collection of **clinical skin disease images** designed for benchmarking computer vision models in dermatological diagnosis. It contains **6,584 images** across **198 different disease categories**, making it one of the most comprehensive publicly available datasets for skin disease classification. A subset, **SD-128**, includes only classes with more than 20 samples to mitigate class imbalance.

## Structure
- **Full dataset (SD-198):** 6,584 images  
- **Subset (SD-128):** Classes with >20 images  
- **Number of classes:** 198 total  
- **Data source:** Images were collected from **DermQuest**, an online dermatology resource where images are submitted by patients and dermatologists.  
- **Annotations:** Each image is labeled and verified by dermatology experts, with additional review by two professionals to ensure quality.

The main folder dataset contains a subfolder and two text files: images, where the classes and images are stored, and the train and test text files which contains subdirectories of the images that are used in trianing and testing.

## Classes:
Extracted classes:
- Acne_Keloidalis_Nuchae  
- Acne_Vulgaris  
- Actinic_solar_Damage(Actinic_Keratosis)  
- Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)  
- Actinic_solar_Damage(Pigmentation)  
- Actinic_solar_Damage(Solar_Elastosis)  
- Actinic_solar_Damage(Solar_Purpura)  
- Actinic_solar_Damage(Telangiectasia)  
- Allergic_Contact_Dermatitis  
- Alopecia_Areata  
- Androgenetic_Alopecia  
- Angioma  
- Angular_Cheilitis  
- Apocrine_Hydrocystoma  
- Basal_Cell_Carcinoma  
- Beau's_Lines  
- Benign_Keratosis  
- Blue_Nevus  
- Bowen's_Disease  
- Callus  
- Candidiasis  
- Cellulitis  
- Clubbing_of_Fingers  
- Compound_Nevus  
- Congenital_Nevus  
- Crowe's_Sign  
- Cutaneous_Horn  
- Darier-White_Disease  
- Dermatofibroma  
- Dermatosis_Papulosa_Nigra  
- Desquamation  
- Digital_Fibroma  
- Dilated_Pore_of_Winer  
- Disseminated_Actinic_Porokeratosis  
- Drug_Eruption  
- Dry_Skin_Eczema  
- Dyshidrosiform_Eczema  
- Dysplastic_Nevus  
- Eczema  
- Epidermal_Nevus  
- Epidermoid_Cyst  
- Erythema_Craquele  
- Erythema_Multiforme  
- Exfoliative_Erythroderma  
- Factitial_Dermatitis  
- Favre_Racouchot  
- Fibroma_Molle  
- Fixed_Drug_Eruption  
- Follicular_Mucinosis  
- Granulation_Tissue  
- Granuloma_Annulare  
- Guttate_Psoriasis  
- Halo_Nevus  
- Herpes_Simplex_Virus  
- Herpes_Zoster  
- Hidradenitis_Suppurativa  
- Hyperkeratosis_Palmaris_Et_Plantaris  
- Hypertrichosis  
- Ichthyosis  
- Infantile_Atopic_Dermatitis  
- Inverse_Psoriasis  
- Junction_Nevus  
- Keloid  
- Keratoacanthoma  
- Keratolysis_Exfoliativa_of_Wende  
- Keratosis_Pilaris  
- Kerion  
- Koilonychia  
- Lentigo_Maligna_Melanoma  
- Leukocytoclastic_Vasculitis  
- Leukonychia  
- Lichen_Planus  
- Lichen_Sclerosis_Et_Atrophicus  
- Lichen_Simplex_Chronicus  
- Lipoma  
- Livedo_Reticularis  
- Lymphocytic_Infiltrate_of_Jessner  
- Malignant_Melanoma  
- Median_Nail_Dystrophy  
- Metastatic_Carcinoma  
- Milia  
- Myxoid_Cyst  
- Nail_Dystrophy  
- Nail_Psoriasis  
- Neurodermatitis  
- Neurofibroma  
- Nevus_Incipiens  
- Nevus_Sebaceous_of_Jadassohn  
- Nevus_Spilus  
- Nummular_Eczema  
- Onycholysis  
- Onychomycosis  
- Onychoschizia  
- Paronychia  
- Perioral_Dermatitis  
- Pitted_Keratolysis  
- Pityriasis_Rosea  
- Pityrosporum_Folliculitis  
- Pseudofolliculitis_Barbae  
- Pseudorhinophyma  
- Psoriasis  
- Pustular_Psoriasis  
- Pyogenic_Granuloma  
- Radiodermatitis  
- Rhinophyma  
- Rosacea  
- Scalp_Psoriasis  
- Scar  
- Sebaceous_Gland_Hyperplasia  
- Seborrheic_Dermatitis  
- Seborrheic_Keratosis  
- Skin_Tag  
- Stasis_Dermatitis  
- Stasis_Edema  
- Stasis_Ulcer  
- Steroid_Acne  
- Steroid_Use_abusemisuse_Dermatitis  
- Striae  
- Syringoma  
- Tinea_Corporis  
- Tinea_Cruris  
- Tinea_Faciale  
- Tinea_Manus  
- Tinea_Pedis  
- Tinea_Versicolor  
- Ulcer  
- Verruca_Vulgaris  
- Xerosis

# Explanatory Data Analaysis (EDA)

Before starting on creating our model, let us first learn about the dataset we are dealing with.  
Through this EDA, there are three main goals we aim to achieve:

1. **Understand the dataset structure**  
   - Verify the number of images and classes.  
   - Explore how the dataset is organized (folders, train/test splits, annotations).  
   - Identify potential data quality issues such as missing or corrupted files.

2. **Analyze class distribution and balance**  
   - Examine how many samples exist per disease category.  
   - Detect any imbalance between common and rare classes.  
   - Visualize the overall image count distribution across all 198 classes.

3. **Inspect image characteristics**  
   - Explore image dimensions, aspect ratios, and resolutions.  
   - Visualize sample images from multiple classes to understand diversity.  
   - Assess potential challenges such as intra-class variation, fine-grained distinctions, and inconsistent lighting or quality.

By performing this EDA, we will gain a better understanding of the **SD-198 dataset's diversity and challenges**, which will guide us in **designing preprocessing steps** and **selecting appropriate modeling strategies**.


## Understanding the Dataset Structure

### Loading the Data
Let us first unzip the dataset and load the data into a pandas dataframe but before that, we need to import the basic python libraries involved such as pandas and matplotpy for data handling and visualization.

In [ ]:
# Unzip the dataset folder.
"""
Drive structure should be the following:
MyDrive
|-- Colab Notebooks
    |-- CIVI100
        |-- SD-198.zip

Note: select "[N]one when asked to rename.
"""
!unzip "/content/drive/MyDrive/Colab Notebooks/CIVI100/SD-128.zip"

Archive:  /content/drive/MyDrive/Colab Notebooks/CIVI100/SD-128.zip
replace __MACOSX/._SD-128? [y]es, [n]o, [A]ll, [N]one, [r]ename: N


In [ ]:
# Basic data handling
import os
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Image handling
from PIL import Image
from tqdm import tqdm

In [ ]:
data_dir = '/content/SD-128/images'

image_paths = []
classes = []

# Get class names from the directories
class_names = [d for d in os.listdir(data_dir) if os.path.isdir(os.path.join(data_dir, d))]

# Collect all image paths and assign classes by walking through the directories
for class_name in class_names:
    class_dir = os.path.join(data_dir, class_name)
    if os.path.isdir(class_dir):
        for image_name in os.listdir(class_dir):
            if image_name.endswith(('.jpg', '.jpeg', '.png')): # Add other image extensions if necessary
                image_full_path = os.path.join(class_dir, image_name)
                image_paths.append(image_full_path)
                classes.append(class_name)

# Create a DataFrame
df = pd.DataFrame({'image_path': image_paths, 'class': classes})

display(df.head())

image_path  \
0  /content/SD-128/images/Actinic_solar_Damage(Cu...   
1  /content/SD-128/images/Actinic_solar_Damage(Cu...   
2  /content/SD-128/images/Actinic_solar_Damage(Cu...   
3  /content/SD-128/images/Actinic_solar_Damage(Cu...   
4  /content/SD-128/images/Actinic_solar_Damage(Cu...   

                                             class  
0  Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)  
1  Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)  
2  Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)  
3  Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)  
4  Actinic_solar_Damage(Cutis_Rhomboidalis_Nuchae)

In [ ]:
print("Total images:", len(df))
print("Total classes:", df['class'].nunique())
print("\nAverage images per class:", len(df)/df['class'].nunique())

df['class'].value_counts().describe()

Total images: 5619
Total classes: 128

Average images per class: 43.8984375


count    128.000000
mean      43.898438
std       15.420020
min       20.000000
25%       28.000000
50%       46.000000
75%       60.000000
max       60.000000
Name: count, dtype: float64

Thanks to that short surface-level description of the dataset, we cna confirm and concur a few things:

- **Total images:** This confirms that the DataFrame contains 5619 images.
- **Total classes:** This shows that there are 128 unique disease categories in the SD-128 subset.
- **Average images per class:** This indicates that, on average, each class has about 43.9 images.
- **Descriptive statistics of class counts:** The `describe()` output for `df['class'].value_counts()` gives a more detailed picture of the class distribution:
    - **count:** There are 128 classes represented in the statistics.
    - **mean:** The average number of images per class is 43.89.
    - **std:** The standard deviation of 15.42 shows the spread of image counts across classes.
    - **min:** The minimum number of images in any class is 20, which aligns with the description of the SD-128 subset only including classes with more than 20 samples.
    - **25%, 50% (median), 75%:** These percentiles show the distribution of class sizes. For example, 50% of the classes have 46 or fewer images.
    - **max:** The maximum number of images in any single class is 60.

These statistics help us understand the size and class balance of the SD-128 dataset, which is an important step in the EDA process.

## Analyzing the Class Distribution and Balance